File: learn-scala-sample/src/main/scala/com/jwd/java/sample/recursion/Fibonacci.java
package com.jwd.java.sample.recursion;

import java.math.BigInteger;

public class Fibonacci {
	
	public static void printFib(int n, BigInteger a, BigInteger b) {
		if (n == 0) {
			return;
		} else {
			System.out.println(n + " - " + a.add(b));
			printFib(n - 1, b, a.add(b));
		}
	}	
	
	public static void main(String args[]) {
		printFib(10, new BigInteger("-1"), BigInteger.ONE);
	}
}

File: learn-scala-sample/src/main/scala/com/jwd/scala/sample/recursion/Fibonacci.scala
package com.jwd.scala.sample.recursion

import scala.annotation.tailrec

object Fibonacci extends App {
	def fib3( n : Int) : Unit = { 
	  @tailrec
	  def fib_tail( n: Int, a:BigInt, b:BigInt): BigInt = {
	    printf("%5d - %d\n", n, a+b);
		n match {
	  		case 1 => a+b 
	  		case _ => fib_tail( n-1, b, a+b )
	    }
	  }
	  
	  fib_tail( n, -1, 1)
	}
	
	fib3(10);
}